File: proj3/grain.go
package main

import (
	"fmt"
	"math"
	"math/rand"
)

type state struct {
	Year  int
	Month int

	Precip  float64
	Temp    float64
	Height  float64
	NumDeer int
}

// can change starting values
var now = &state{
	Year:    2019,
	Month:   0,
	Height:  1,
	NumDeer: 1,
}

// plus fourth something

// can change if we want
const (
	grainGrowsPerMonth  = 8.0 // inches
	oneDeerEatsPerMonth = 0.5

	avgPrecipPerMonth = 6.0
	ampPrecipPerMonth = 6.0
	randomPrecip      = 2.0

	avgTemp    = 50.0
	ampTemp    = 20.0
	randomTemp = 10.0

	midTemp   = 40.0
	midPrecip = 10.0
)

func main() {
	updateTemperatureAndPrecipitation(now)
	//printResults()
	deerInC := make(chan *state, 1)
	grainInC := make(chan *state, 1)

	deerC := make(chan int, 1)
	grainC := make(chan float64, 1)

	go GrainDeer(deerInC, deerC)
	go GrainGrowth(grainInC, grainC)
	go MyAgent()

	for now.Year < 2025 {
		deerInC <- now
		grainInC <- now

		nextHeight := <-grainC
		nextDeer := <-deerC

		now.Height = nextHeight
		now.NumDeer = nextDeer

		//print results and increment time
		printResults()
		updateYear(now)
		updateTemperatureAndPrecipitation(now)
	}
}

func printResults() {
	fmt.Println(now.Year, now.Month, now.Temp, now.Precip, now.NumDeer, now.Height)
}

func updateYear(now *state) {
	now.Month++
	if now.Month == 12 {
		now.Year++
		now.Month = 0
	}
}

var rng = rand.New(rand.NewSource(0))

func updateTemperatureAndPrecipitation(now *state) {
	ang := float64(30*now.Month+15) * (math.Pi / 180)

	temp := avgTemp - ampTemp*math.Cos(ang)
	now.Temp = temp + ranf(rng, -randomTemp, randomTemp)

	precip := avgPrecipPerMonth + ampPrecipPerMonth*math.Sin(ang)
	now.Precip = precip + ranf(rng, -randomPrecip, randomPrecip)
	if now.Precip < 0 {
		now.Precip = 0
	}
}

func GrainDeer(in <-chan *state, out chan<- int) {
	//  The Carrying Capacity of the graindeer is the number of inches of height of the grain. If the number of graindeer exceeds this value at the end of a month, decrease the number of graindeer by one. If the number of graindeer is less than this value at the end of a month, increase the number of graindeer by one.
	//
	for {
		now := <-in
		carryingCapacity := now.Height
		nextNumDeer := now.NumDeer

		if float64(now.NumDeer) > carryingCapacity {
			nextNumDeer -= 1
		} else if float64(now.NumDeer) < carryingCapacity {
			nextNumDeer += 1
		}

		out <- nextNumDeer
	}
}

func GrainGrowth(in <-chan *state, out chan<- float64) {
	//  Each month you will need to figure out how much the grain grows. If conditions are good, it will grow by GRAIN_GROWS_PER_MONTH. If conditions are not good, it won't.
	//
	//  You know how good conditions are by seeing how close you are to an ideal temperature (°F) and precipitation (inches). Do this by computing a Temperature Factor and a Precipitation Factor like this:

	for {
		now := <-in
		// this function peaks at 1 around midTemp, and peters off to 0 at plus or minus 10
		tempFactor := math.Exp(-square((now.Temp - midTemp) / 10))
		precipFactor := math.Exp(-square((now.Precip - midPrecip) / 10))

		nextHeight := now.Height
		nextHeight += tempFactor * precipFactor * grainGrowsPerMonth
		nextHeight -= float64(now.NumDeer) * oneDeerEatsPerMonth
		if nextHeight < 0 {
			nextHeight = 0
		}

		out <- nextHeight
	}
}

func MyAgent() {}

func square(x float64) float64 { return x * x }

func ranf(rng *rand.Rand, low, high float64) float64 {
	return rng.Float64()*(high-low) - low
}

File: proj3/grain.cpp
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <omp.h>

long nowYear  = 2019;
long nowMonth = 0;
	// can change starting values

double nowPrecip ;
double nowTemp   ;
double nowHeight = 1.0;
long nowNumDeer = 1;
double nowMystery = 1.0;


// plus fourth something

// can change if we want
const long maxYear = 2025;
const double grainGrowsPerMonth  = 8.0; // inches
const double oneDeerEatsPerMonth = 0.5;
const double numMonthsForGrainDensityToDouble = 18.0;

const double avgPrecipPerMonth = 6.0;
const double ampPrecipPerMonth = 6.0;
const double randomPrecip      = 2.0;

const double avgTemp    = 50.0;
const double ampTemp    = 20.0;
const double randomTemp = 10.0;

const double midTemp   = 40.0;
const double midPrecip = 10.0;

void Watcher();
void GrainDeer();
void GrainGrowth();
void MyAgent();
void updateTemperatureAndPrecipitation();
void printResults();
void updateYear();
double Ranf( unsigned int *seedp,  double low, double high );
void barrier();

static double square(double x ) { return x * x; }

int main() {
	omp_set_num_threads(4);
	updateTemperatureAndPrecipitation();
	#pragma omp parallel sections
	{
		#pragma omp section
		Watcher();
		#pragma omp section
		GrainDeer();
		#pragma omp section
		GrainGrowth();
		#pragma omp section
		MyAgent();
	}
	return 0;
}

void Watcher() {
	while (nowYear < maxYear) {
		// compute a temporary next-value for this quantity
		// based on the current state of the simulation:
		//. . .

		// DoneComputing barrier:
		barrier();

		// DoneAssigning barrier:
		barrier();

		//print results and increment time
		printResults();
		updateYear();
		updateTemperatureAndPrecipitation();

		// DonePrinting barrier:
		barrier();
	}

}

void printResults() {
	double temp = (nowTemp - 32.) * (5./9.); // convert to celcius
	printf("%ld %ld %f %f %ld %f %f\n", nowYear, nowMonth, temp, nowPrecip, nowNumDeer, nowHeight, nowMystery);
}

void updateYear() {
	nowMonth++;
	if (nowMonth == 12) {
		nowYear++;
		nowMonth = 0;
	}
}

void updateTemperatureAndPrecipitation() {
	static unsigned int seed = 0;
	double ang = ((double)(30.*nowMonth+15.)) * (M_PI / 180.);

	double temp = avgTemp - ampTemp*cos(ang);
	nowTemp = temp + Ranf(&seed, -randomTemp, randomTemp);

	double precip = avgPrecipPerMonth + ampPrecipPerMonth*sin(ang);
	nowPrecip = precip + Ranf(&seed, -randomPrecip, randomPrecip);
	if (nowPrecip < 0.) {
		nowPrecip = 0.;
	}
}

void GrainDeer() {
	//  The Carrying Capacity of the graindeer is the number of inches of height of the grain. If the number of graindeer exceeds this value at the end of a month, decrease the number of graindeer by one. If the number of graindeer is less than this value at the end of a month, increase the number of graindeer by one.
	//
	while (nowYear < maxYear) {
		double carryingCapacity = nowHeight;
		long nextNumDeer = nowNumDeer;

		if (((double)(nowNumDeer)) > carryingCapacity) {
			nextNumDeer -= 1;
		} else if (((double)(nowNumDeer)) < carryingCapacity) {
			nextNumDeer += nowMystery;
		}

		barrier();

		nowNumDeer = nextNumDeer;

		barrier();
		barrier();
	}
}

void GrainGrowth() {
	//  Each month you will need to figure out how much the grain grows. If conditions are good, it will grow by GRAIN_GROWS_PER_MONTH. If conditions are not good, it won't.
	//
	//  You know how good conditions are by seeing how close you are to an ideal temperature (°F) and precipitation (inches). Do this by computing a Temperature Factor and a Precipitation Factor like this:

	while (nowYear < maxYear) {
		// this function peaks at 1 around midTemp, and peters off to 0 at plus or minus 10
		double tempFactor = exp(-square((nowTemp - midTemp) / 10));
		double precipFactor = exp(-square((nowPrecip - midPrecip) / 10));

		double nextHeight = nowHeight;
		nextHeight += tempFactor * precipFactor * grainGrowsPerMonth;
		nextHeight -= ((double)(nowNumDeer)) * oneDeerEatsPerMonth;
		if (nextHeight < 0) {
			nextHeight = 0;
		}

		barrier();
		nowHeight = nextHeight;
		barrier();
		barrier();
	}
}

/* ideas:
 * - grain strain (a certain percentage of grain is poisonous to graindeer
 * - farmer cuts grain every so often
 * - wolfs which eat deer
 * - drought in a certain year
 * - bugs which eat the grain too
 * - another type of grain
 * - periodic fires
 */
void MyAgent() {
	// Approximately every 18 months, grain density doubles: the grain
	// farmers are able to grow more grain in the same amount of space.
	// This has the side effect of promoting graindeer growth because they
	// can now raise more children because they don't have to spend as much
	// time searching for food. The number of additional graindeer each
	// month (assuming the grain is of a sufficient height) is precisely
	// proportional to the grain density.
	//
	double densityGrowthPerMonth = pow(2.0, 1./numMonthsForGrainDensityToDouble);
	while (nowYear < maxYear) {
		double nextMystery = nowMystery * densityGrowthPerMonth;

		barrier();

		nowMystery = nextMystery;

		barrier();
		barrier();
	}
}


double Ranf( unsigned int *seedp,  double low, double high )
{
	double r = (double) rand_r(seedp);  // 0 - RAND_MAX
	return low  +  r * ( high - low ) / (double)RAND_MAX;
}

void barrier() {
	#pragma omp barrier
}
